Sources/_MatchingEngine/Regex/AST/Atom.swift

```diff
@@ -382,35 +382,37 @@ extension AST.Atom.CharacterProperty {
   }
 }
 
+extension AST.Atom {
+  public struct Reference: Hashable {
+    @frozen
+    public enum Kind: Hashable {
+      // \n \gn \g{n} \g<n> \g'n' (?n) (?(n)...
+      // Oniguruma: \k<n>, \k'n'
+      case absolute(Int)
+
+      // \g{-n} \g<+n> \g'+n' \g<-n> \g'-n' (?+n) (?-n)
+      // (?(+n)... (?(-n)...
+      // Oniguruma: \k<-n> \k<+n> \k'-n' \k'+n'
+      case relative(Int)
+
+      // \k<name> \k'name' \g{name} \k{name} (?P=name)
+      // \g<name> \g'name' (?&name) (?P>name)
+      // (?(<name>)... (?('name')... (?(name)...
+      case named(String)
+
+      // ?(R) (?(R)...
+      case recurseWholePattern
+    }
+    public var kind: Kind
+    public var location: SourceLocation
```

Comment:
Would this be different than `Atom`'s location? Or does this track the number/name inside the parens/escape? If the latter, then I think we need a name and comment.